src/flamegraph/color.rs

```diff
@@ -0,0 +1,272 @@
+        Palette::Blue => return ((80, 60, v1), (80, 60, v1), (205, 60, v1)),
+        Palette::Yellow => return ((175, 55, v1), (175, 55, v1), (50, 20, v1)),
+        Palette::Purple => return ((190, 65, v1), (80, 60, v1), (190, 65, v1)),
+        Palette::Aqua => return ((50, 60, v1), (165, 55, v1), (165, 55, v1)),
+        Palette::Orange => return ((190, 65, v1), (90, 65, v1), (0, 0, v1)),
+        Palette::Java => handle_java_palette(name),
+        Palette::Perl => handle_perl_palette(name),
+        Palette::Js => handle_js_palette(name),
+        Palette::Wakeup => handle_wakeup_palette(name),
+        Palette::Chain => handle_chain_palette(name),
+    };
+
+    coefficients_for_palette(&real_palette, name, v1, v2, v3)
+}
+
+fn affine_transform(a: u8, b: u8, x: f32) -> u8 {
+    b + (a as f32 * x) as u8
+}
+
+fn color_from_palette(palette: &Palette, name: &str, v1: f32, v2: f32, v3: f32) -> String {
+    let ((r_b, r_a, r_x),
+         (g_b, g_a, g_x),
+         (b_b, b_a, b_x)) = coefficients_for_palette(palette, name, v1, v2, v3);
+
+    let red = affine_transform(r_a, r_b, r_x);
+    let green = affine_transform(g_a, g_b, g_x);
+    let blue = affine_transform(b_a, b_b, b_x);
+
+    format!("rgb({},{},{})", red, green, blue)
+}
+
+pub(super) fn color_map<'a>(palette: &Palette, hash: bool, name: &'a str, palette_map: &'a mut HashMap<String, String>) -> &'a str {
+    palette_map.entry(name.to_string()).or_insert_with(|| color(palette, hash, name))
+}
+
+pub(super) fn color(palette: &Palette, hash: bool, name: &str) -> String {
+    let (v1, v2, v3) = if hash {
+        let name_hash = namehash(name);
+        let reverse_name_hash = namehash(&name.chars().rev().collect::<String>());
+
+        (name_hash, reverse_name_hash, reverse_name_hash)
+    } else {
+        (rand::random(), rand::random(), rand::random())
+    };
+
+    color_from_palette(palette, name, v1, v2, v3)
+}
+
+pub(super) fn get_background_colors_for(palette: &Palette) -> (&str, &str) {
+    match palette {
+        Palette::Hot | Palette::Java | Palette::Js | Palette::Perl => YELLOW_GRADIENT,
+        Palette::Mem | Palette::Chain => BLUE_GRADIENT,
+        _ => GRAY_GRADIENT
+    }
+}
+
+pub(super) fn read_palette(file: &str) -> Result<HashMap<String, String>, io::Error> {
+    let mut map = HashMap::default();
+    let path = Path::new(file);
+
+    if path.exists() {
+        let file = File::open(path)?;
+        let file = BufReader::new(file);
+
+        for line in file.lines() {
+            let line = line?;
+            let words = line.split("->").collect::<Vec<_>>();
+            map.insert(words[0].to_string(), words[1].to_string());
+        }
+    }
+
+    Ok(map)
+}
+
+pub(super) fn write_palette(file: &str, palette_map: HashMap<String, String>) -> Result<(), io::Error> {
+    let mut file = OpenOptions::new().write(true).create(true).open(file)?;
+    let mut entries = palette_map.into_iter().collect::<Vec<_>>();
+    entries.sort_unstable();
```

Comment:
Is it important that these are sorted?

Reply:
I guess the Perl version [does sort](https://github.com/brendangregg/FlameGraph/blob/f857ebc94bfe2a9bfdc4f1536ebacfb7466f69ba/flamegraph.pl#L514), but I don't quite see why?

Reply:
I don't know if other tools may use the `palette.map` file ?
Sorting is nice if we want to compare with the `flamegraph.pl` output.

Reply:
Yeah, that's a good point. Let's keep it this way then.

Reply:
Actually, can you put in a comment saying "We sort the palette because the Perl implementation does"?

Reply:
[Done](https://github.com/jonhoo/inferno/pull/32/commits/b0a98a8940220af42cdd3d9b2a74a506863cb179)